### PR TITLE
Fix incorrect string to type conversion

### LINKED ISSUE #19559: `[type]` cast and `-as [Type]` malfunction with strings that contain "]" (only)
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```powershell
# OK
[Type] 'int'
'int' -as [Type]

# !! BROKEN - these calls succeed, despite obviously broken input.
# Note that prepending "[" *would* fail, as expected, given that [Type] "[int]" doesn't - and shouldn't - work.
# It doesn't matter what, if anything, follows the closing-only "]"
[Type] 'int]whatever'
'int]whatever' -as [Type]
```

### Expected behavior

Output of two `[type]` instances, and two errors.

### Actual behavior

Output of _four_ `[int]` instances. That is, the broken type names were unexpectedly accepted.

### Error details

_No response_

### Environment data

```powershell
PowerShell Core 7.4.0-preview.3
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes incorrect string to type conversions like: `[Type] 'int]whatever'` so they are no longer parsed as types.  
This was fixed by adding a check for `ScanType` to also check that the parsed type name matches the input type name.

## PR Context
Fixes https://github.com/PowerShell/PowerShell/issues/19559

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Language/Parser/Conversions.Tests.ps1 [hunk 1]
```diff
@@ -78,6 +78,10 @@ Describe 'conversion syntax' -Tags "CI" {
             $result -join ";" | Should -Be ($Elements -join ";")
         }
     }
+
+    It 'Should not convert invalid strings to type name' {
+        'int]whatever' -as [type] | Should -Be $null
```
Comment (andschwa) [inline test/powershell/Language/Parser/Conversions.Tests.ps1:83]: Suggesting to add the other report from the issue.
```suggestion
        'int]whatever' -as [type] | Should -Be $null
        [Type] 'int]whatever' | Should -Be $null
```

**Review by andyleejordan — APPROVED** [review]
Good fix! Though @SeeminglyScience do you think it should error instead of returning null? I'm unsure what's more correct.

File: test/powershell/Language/Parser/Conversions.Tests.ps1 (hunk 1 above)
Comment (andschwa) [reply test/powershell/Language/Parser/Conversions.Tests.ps1:83]: Err, so checked with @SeeminglyScience and while the former should be null, the latter cast should throw an invalid argument error.

File: test/powershell/Language/Parser/Conversions.Tests.ps1 [hunk 2]
```diff
@@ -78,6 +78,10 @@ Describe 'conversion syntax' -Tags "CI" {
...
+    }
```
Comment (andschwa) [inline test/powershell/Language/Parser/Conversions.Tests.ps1:84]:
```suggestion
    }
    
    It 'Should not convert invalid strings to type name using left-hand side operator' {
        [Type] 'int]whatever' | Should -Throw
    }
```

File: test/powershell/Language/Parser/Conversions.Tests.ps1 [hunk 3]
```diff
@@ -78,6 +78,10 @@ Describe 'conversion syntax' -Tags "CI" {
             $result -join ";" | Should -Be ($Elements -join ";")
         }
     }
+
+    It 'Should not convert invalid strings to type name' {
```
Comment (andschwa) [inline test/powershell/Language/Parser/Conversions.Tests.ps1:82]:
```suggestion
    It 'Should not convert invalid strings to type name using -as operator' {
```

**Review by andyleejordan — CHANGES_REQUESTED** [review]
New suggestion of a test using the left-hand side operator (which _should_ throw now like with other errors, and I don't if it will).

File: test/powershell/Language/Parser/Conversions.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [reply test/powershell/Language/Parser/Conversions.Tests.ps1:84]: What is `-ErrorId` we expect?

Comment (andschwa) [reply test/powershell/Language/Parser/Conversions.Tests.ps1:84]: Probably`InvalidArgument` based on this experiment:

```
> [Type] 'int]whatever'

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Int32                                    System.ValueType

@andys-mac-mini ~ 
> [Type] 'intwhatever' 
InvalidArgument: Cannot convert the "intwhatever" value of type "System.String" to type "System.Type".
```

Comment (iSazonov) [reply test/powershell/Language/Parser/Conversions.Tests.ps1:84]: I mean we usually use `-ErrorId` in `Should -Throw`.

Comment (andschwa) [reply test/powershell/Language/Parser/Conversions.Tests.ps1:84]: Ah, well I don't know 🙃

**Review by andyleejordan — APPROVED** [review]
Ah, I guess it needed braces 💁

**Pull request merged**